### DISCUSSION TITLE
How do I specify `Index` for node's ancestor in `visitParents`?

### DISCUSSION THREAD

**Discussion by shxxyz** [syntax-tree/unist#126] [Q&A] [answered]
I add an element to an ancestor of node and `visitParents` keeps finding same element.

```yaml
- root
  - ancestor
    - ...more ancestors
      - parent
        - this is the element
```

And I add an element to its ancestor:

```yaml
- root
  - ancestor
    - this is what I add
    - ...more ancestors
      - parent
        - this is the element
```

How do I return from visitor so that, next time it's traversing `ancestor`, it skip both `this is what I add` and `...more ancestors`?

there's an ugly solution tho, record every valid element in `TestFunction`, and if the element was found before, then return `false`, but I think this is not the right way.

**Comment by wooorm** [comment]
This isn’t made for that, indeed, you are not allowed to do that.

What I’d recommend is to track a list of `Array<[parent: Parent, child: Node, replacement: Node]>` in the visitor.
And then *after* walking the entire tree, *then* you do:

```
for ([parent, child, replacement] of stack) {
  index = parent.children.indexOf(child)
  parent.children[index] = replacement
  
}
```

But perhaps this is an XY question. What and why do you want to do this?

**Reply by shxxyz** [reply]
Actually, I want to show footnotes near the content, but I am too lazy to write a footnote parser in my own, so I decided to use `remark-gfm`. Therefore I am writing a rehype plugin to make a copy of footnote elements to where it is referenced(then do some fancy css trics).  
The footnote is copied to the reference's lowest `<p>` ancestor's parent, to find that ancestor and its parent, I think there's no better tool than `unist-util-visit-parent`.

**Reply by wooorm** [reply, marked as answer]
`remark-gfm` is unrelated to how those footnotes are “displayed” in HTML. It’s a parser. If you want to display footnotes in, say, some very weird way, that’s possible with `remark-gfm`.

Where HTML comes in, is in `remark-rehype`. *That’s* what decides how those markdown footnotes are turned into HTML.

There are already many options there, relating to footnotes. If those are not enough, from what you are trying to do, I recommend making handlers for `remark-rehype` instead.

---

> The footnote is copied to the reference's lowest `<p>` ancestor's parent, to find that ancestor and its parent, I think there's no better tool than unist-util-visit-parent.

All of this can be true *and* be unrelated to your problem.

It is still vague to me what you want to do. If you post actual input markdown and expected HTML I can better help you.

I think you are asking for `visit` in `visit`.

**Reply by shxxyz** [reply]
Ah, now I see I can customize the handler to do this, but I'll stick with nested `visit` for now, the handler docs on `remark-rehype` is a bit hard to read for a newbie like me, it's like machine generated. I will try to write a custom handler if I have time reading that. Thanks!
